pallets/order-book/src/order_book.rs

```diff
@@ -108,4 +119,168 @@ impl<T: crate::Config + Sized> OrderBook<T> {
         self.last_order_id = self.last_order_id.add(T::OrderId::one());
         self.last_order_id
     }
+
+    pub fn place_limit_order<Locker>(
+        &self,
+        order: LimitOrder<T>,
+        data: &mut impl DataLayer<T>,
+    ) -> Result<(), DispatchError>
+    where
+        Locker: CurrencyLocker<T::AccountId, T::AssetId, T::DEXId>,
+    {
+        ensure!(
+            self.status == OrderBookStatus::Trade || self.status == OrderBookStatus::PlaceAndCancel,
+            Error::<T>::PlacementOfLimitOrdersIsForbidden
+        );
+
+        self.ensure_limit_order_valid(&order)?;
+        self.check_restrictions(&order, data)?;
+
+        let cross_spread = match order.side {
+            PriceVariant::Buy => {
+                if let Some((best_ask_price, _)) = self.best_ask(data) {
+                    order.price >= best_ask_price
+                } else {
+                    false
+                }
+            }
+            PriceVariant::Sell => {
+                if let Some((best_bid_price, _)) = self.best_bid(data) {
+                    order.price <= best_bid_price
+                } else {
+                    false
+                }
+            }
+        };
+
+        if cross_spread {
+            if self.status == OrderBookStatus::Trade {
+                self.cross_spread();
+            } else {
+                return Err(Error::<T>::InvalidLimitOrderPrice.into());
+            }
+        }
+
+        let (lock_asset, lock_amount) = order.appropriate_asset_and_amount(&self.order_book_id)?;
+
+        Locker::lock_liquidity(
+            self.dex_id,
+            &order.owner,
+            self.order_book_id,
+            lock_asset,
+            lock_amount,
+        )?;
+
+        data.insert_limit_order(&self.order_book_id, order)?;
+        Ok(())
+    }
+
+    fn ensure_limit_order_valid(&self, order: &LimitOrder<T>) -> Result<(), DispatchError> {
+        order.ensure_valid()?;
+        ensure!(
+            order.price % self.tick_size == 0,
+            Error::<T>::InvalidLimitOrderPrice
+        );
+        ensure!(
+            self.min_lot_size <= order.amount && order.amount <= self.max_lot_size,
+            Error::<T>::InvalidOrderAmount
+        );
+        ensure!(
+            order.amount % self.step_lot_size == 0,
+            Error::<T>::InvalidOrderAmount
+        );
+        Ok(())
+    }
+
+    fn check_restrictions(
+        &self,
+        order: &LimitOrder<T>,
+        data: &mut impl DataLayer<T>,
+    ) -> Result<(), DispatchError> {
+        if let Some(user_orders) = data.get_user_limit_orders(&order.owner, &self.order_book_id) {
+            ensure!(
+                !user_orders.is_full(),
```

Comment:
You could return this error if `try_push` call failed. The same for another checks in this function

Reply:
The idea is to check it before trying to push